rules_cc/defs.bzl

```diff
@@ -12,41 +12,77 @@
-        "@platforms//cpu:x86_64": ["@gcc_toolchain_x86_64//:libstdcxx{}".format("_static" if static_libstdcxx else "")],
+        "@platforms//cpu:aarch64": ["{}_aarch64//:{}".format(gcc_toolchain_prefix, target_name)],
+        "@platforms//cpu:armv7": ["{}_armv7//:{}".format(gcc_toolchain_prefix, target_name)],
+        "@platforms//cpu:x86_64": ["{}_x86_64//:{}".format(gcc_toolchain_prefix, target_name)],
     })
-    native.cc_library(
+    rule(
         name = name,
         deps = deps,
         **kwargs
     )
 
+def cc_library(name, static_libstdcxx = False, **kwargs):
+    _with_libstdcxx(
+        rule = native.cc_library,
+        name = name,
+        gcc_toolchain_prefix = "@gcc_toolchain",
+        static_libstdcxx = static_libstdcxx,
+        **kwargs,
+    )
+
 def cc_binary(name, static_libstdcxx = False, **kwargs):
-    deps = kwargs.pop("deps", [])
-    deps += select({
-        "@platforms//cpu:aarch64": ["@gcc_toolchain_aarch64//:libstdcxx{}".format("_static" if static_libstdcxx else "")],
-        "@platforms//cpu:armv7": ["@gcc_toolchain_armv7//:libstdcxx{}".format("_static" if static_libstdcxx else "")],
-        "@platforms//cpu:x86_64": ["@gcc_toolchain_x86_64//:libstdcxx{}".format("_static" if static_libstdcxx else "")],
-    })
-    native.cc_binary(
+    _with_libstdcxx(
+        rule = native.cc_binary,
         name = name,
-        deps = deps,
-        **kwargs
+        gcc_toolchain_prefix = "@gcc_toolchain",
+        static_libstdcxx = static_libstdcxx,
+        **kwargs,
     )
 
 def cc_test(name, static_libstdcxx = False, **kwargs):
-    deps = kwargs.pop("deps", [])
-    deps += select({
-        "@platforms//cpu:aarch64": ["@gcc_toolchain_aarch64//:libstdcxx{}".format("_static" if static_libstdcxx else "")],
-        "@platforms//cpu:armv7": ["@gcc_toolchain_armv7//:libstdcxx{}".format("_static" if static_libstdcxx else "")],
-        "@platforms//cpu:x86_64": ["@gcc_toolchain_x86_64//:libstdcxx{}".format("_static" if static_libstdcxx else "")],
-    })
-    native.cc_test(
+    _with_libstdcxx(
+        rule = native.cc_test,
         name = name,
-        deps = deps,
-        **kwargs
+        gcc_toolchain_prefix = "@gcc_toolchain",
+        static_libstdcxx = static_libstdcxx,
+        **kwargs,
+    )
+
+# These targets should only be used to patch dependencies of gcc-toolchain that should use our toolchain.
+# For instance, we depend on protobuf to run `examples/protobuf`.
+# However, we also want protobuf to link against libstdc++, so we patch it to use our version of the targets.
+# You may be wondering: Why can't we use the "public" versions of the targets?
+# In Bzlmod, the repositories `@gcc_toolchain_<arch>` that we create in MODULE.bazel are not visible to our dependencies.
+# Therefore, we need to use the "absolute path" of these repositories to address them (main~extension~toolchain.
+# Documentation: https://bazel.build/external/extension#repository_names_and_visibility
```

Comment:
This comment is important to understand why we need a different patch for `protobuf`.